### PR TITLE
Allowing non-contiguous bin edges.

### LINKED ISSUE #3255: Binning with non-contigous edges fails with DimensionError
```python
table = sc.data.table_xyz(1000)
edges = sc.linspace('x', 0, 1, 100, unit='m')
table.bin(x=edges)
```
works, but
```python
table.bin(x=edges[::2])
```
fails with
```python
DimensionError: View over subspan can only be created for contiguous range of data.
```

Expected the binning to work regardless of the contigousness of the edges.

### PR DESCRIPTION
**Pull request by YooSunYoung**
Fixes #3255

### CODE REVIEW

File: lib/dataset/bin_detail.cpp [hunk 1]
```diff
@@ -4,13 +4,13 @@
 /// @author Simon Heybrock
 #include "scipp/core/element/bin.h"
 #include "scipp/core/element/map_to_bins.h"
+#include <iostream>
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.cpp:7]:
```suggestion
```

File: lib/dataset/bin_detail.cpp [hunk 2]
```diff
@@ -38,7 +45,7 @@ void update_indices_by_binning(Variable &indices, const Variable &key,
         "event-coordinate. Provide an event coordinate or convert the "
         "bin-edge coordinate to a non-edge coordinate.");
   const auto &edge_view =
-      is_bins(edges) ? as_subspan_view(edges) : subspan_view(edges, dim);
+      is_bins(edges) ? as_subspan_view(edges) : to_subspan_view(edges, dim);
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.cpp:48]: Also change `update_indices_by_grouping` below? Or is that not needed?

**Review by SimonHeybrock — COMMENTED** [review]
Looks reasonable. Can you add tests?

File: tests/binning_test.py [hunk 3]
```diff
@@ -944,3 +944,52 @@ def test_group_many_groups(ngroup):
...
+    var = noncontiguous_var.copy()
+    assert var.values.data.contiguous
+    return var
+
+
+def test_noncontiguous_binning(
+    noncontiguous_var: sc.Variable, contiguous_var: sc.Variable
+):
+    table = sc.data.table_xyz(1_000)
+    assert sc.identical(table.bin(x=noncontiguous_var), table.bin(x=contiguous_var))
```
Comment (YooSunYoung) [inline tests/binning_test.py:967]: This test fails, but I don't know why it fails...

Here is the code snippet to reproduce this:
```python
table = sc.data.table_xyz(10_000)
noncont_var = sc.linspace('x', 0, 1, 101, unit='m')[::2]
con_var = noncont_var.copy()
assert sc.identical(noncont_var, con_var)
assert not sc.identical(table.bin(x=noncont_var), table.bin(x=con_var))
```
And the first two bins of each binned data array were different like this:
image


```

File: lib/dataset/bin_detail.cpp [hunk 4]
```diff
@@ -28,17 +28,26 @@ Variable make_range(const scipp::index begin, const scipp::index end,
...
+namespace {
+template <class T> Variable as_edge_view(T &edges) {
+  if (is_bins(edges)) {
+    return as_subspan_view(edges);
+  }
+  const Dim dim = edges.dims().inner();
+  const auto con_edges = scipp::variable::as_contiguous(edges, dim);
+  return subspan_view(con_edges, dim);
+}
+} // namespace
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.cpp:40]: Regarding your failing test: See the big warning in `histogram.cpp` about `as_contiguous` returning a new variable. Aren't you running into the same problem here?

File: lib/dataset/bin_detail.cpp [hunk 5]
```diff
@@ -28,17 +28,31 @@ Variable make_range(const scipp::index begin, const scipp::index end,
...
+  // and kept until the end of the expression where it is called.
+  if (is_bins(edges)) {
+    return std::make_pair(edges, as_subspan_view(edges));
+  }
+  const Dim dim = edges.dims().inner();
+  const auto con_edges = scipp::variable::as_contiguous(edges, dim);
+  return std::make_pair(con_edges, subspan_view(con_edges, dim));
+}
+} // namespace
+
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.cpp:44]: The helper functions seems to just complicate everything. It is used only once, so why not put the code inline, then the headache about the out-of-scope `con_edges` goes away.

File: lib/dataset/histogram.cpp [hunk 6]
```diff
@@ -94,23 +82,25 @@ DataArray histogram(const DataArray &events, const Variable &binEdges) {
...
 Dataset histogram(const Dataset &dataset, const Variable &binEdges) {
+  auto dim = binEdges.dims().inner();
+  auto con_bin_edges = as_contiguous(binEdges, dim);
   return apply_to_items(
       dataset,
       [](const auto &item, const Dim, const auto &binEdges_) {
         return histogram(item, binEdges_);
       },
-      binEdges.dims().inner(), binEdges);
+      dim, con_bin_edges);
```
Comment (SimonHeybrock) [inline lib/dataset/histogram.cpp:103]: Why is this modified?

File: lib/dataset/histogram.cpp [hunk 7]
```diff
@@ -94,23 +82,25 @@ DataArray histogram(const DataArray &events, const Variable &binEdges) {
...
         },
-        event_dim, binEdges);
+        event_dim, con_bin_edges);
   } else {
     throw except::BinEdgeError(
         "Data is already histogrammed. Expected event data or dense point "
         "data, got data with bin edges.");
   }
-  result.coords().set(dim, binEdges);
+  result.coords().set(dim, con_bin_edges);
```
Comment (SimonHeybrock) [inline lib/dataset/histogram.cpp:91]: I think this should keep setting `binEdges` as the output coord?

File: lib/dataset/bin_detail.h [hunk 8]
```diff
@@ -6,16 +6,18 @@
...
 template <class T> Variable as_subspan_view(T &&binned) {
   auto &&[indices, dim, buffer] = binned.template constituents<Variable>();
-  if constexpr (std::is_const_v<std::remove_reference_t<T>>)
+  if constexpr (std::is_const_v<std::remove_reference_t<T>>) {
     return subspan_view(std::as_const(buffer), dim, indices);
-  else
+  } else {
     return subspan_view(buffer, dim, indices);
+  }
 }
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.h:21]: Why was this file modified? Why the new include?

File: lib/dataset/test/bin_test.cpp [hunk 9]
```diff
@@ -707,3 +707,53 @@ TEST(BinLinspaceTest, many_events_many_bins) {
...
+  EXPECT_TRUE(cont_x.stride(Dim::X) == 1);
+
+  const auto x_edges =
+      makeVariable<double>(Dims{Dim::X}, Shape{5}, Values{-2, -1, 0, 1, 2});
+  const scipp::core::Slice slice(Dim::X, 0, 5, 2);
+  const auto non_cont_x = x_edges.slice(slice);
+  EXPECT_TRUE(non_cont_x.stride(Dim::X) != 1);
+
+  EXPECT_EQ(bin(table, {non_cont_x}), bin(table, {cont_x}));
+}
```
Comment (SimonHeybrock) [inline lib/dataset/test/bin_test.cpp:724]: Can you check in all cases that the coord of the input still references the "non-cont" variable provided by the caller, and NOT the copy made internally.

File: lib/dataset/bin_detail.h (hunk 8 above)
Comment (SimonHeybrock) [reply lib/dataset/bin_detail.h:21]: I think there can be obscure ways to get a non-contiguous buffer, but I think we should ignore that, for now at least.

File: lib/dataset/histogram.cpp (hunk 6 above)
Comment (YooSunYoung) [reply lib/dataset/histogram.cpp:103]: Seems like a mistake... I removed it.

File: lib/dataset/bin_detail.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply lib/dataset/bin_detail.cpp:7]: Github still shows the include in the diff.

File: lib/dataset/bin_detail.cpp (hunk 5 above)
Comment (YooSunYoung) [reply lib/dataset/bin_detail.cpp]: I just removed the helper and moved the body of the function but I thought they are around the same since they have to be wrapped in the if ``is_bins`` statement anyways...

File: tests/binning_test.py (hunk 3 above)
Comment (YooSunYoung) [reply tests/binning_test.py:967]: This one is now fixed...!

File: lib/dataset/test/bin_test.cpp (hunk 9 above)
Comment (YooSunYoung) [reply lib/dataset/test/bin_test.cpp:724]: I still couldn't figure out how to access to the original reference...

File: lib/dataset/bin_detail.cpp [hunk 10]
```diff
@@ -30,23 +29,32 @@ Variable make_range(const scipp::index begin, const scipp::index end,
...
-  const auto &edge_view =
-      is_bins(edges) ? as_subspan_view(edges) : subspan_view(edges, dim);
+
+  Variable con_edges;
+  Variable edge_view;
+
+  if (is_bins(edges)) {
+    edge_view = as_subspan_view(edges);
+  } else {
+    const Dim dim = edges.dims().inner();
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.cpp:45]: You can undo the inlining of `edges.dims().inner()` so we don't have the call twice.

File: lib/dataset/bin_detail.h [hunk 11]
```diff
@@ -6,6 +6,7 @@
 
 #include "scipp-dataset_export.h"
 #include "scipp/variable/subspan_view.h"
+#include "scipp/variable/util.h"
```
Comment (SimonHeybrock) [inline lib/dataset/bin_detail.h:9]: unused?

File: lib/dataset/test/bin_test.cpp (hunk 9 above)
Comment (SimonHeybrock) [reply lib/dataset/test/bin_test.cpp:724]: Something like
```python
EXPECT_EQ(&result.coords()[Dim::X].values<double>()[0], x_edges.values<double>()[0])
```
might work.

**Review by SimonHeybrock — APPROVED** [review]

**Comment by YooSunYoung** [comment]
I'll continue with adding tests of reference check after #3323 is merged.
